test/test_composability.py

```diff
@@ -174,6 +184,79 @@ def test_manual_pipeline_with_manual_allreduce(self):
         ddp_pp_model.all_reduce(num_microbatches)
         print(f"{self.rank} finished all_reduce")
 
+    def test_manual_pipeline_with_fsdp(self):
+        device_mesh, device = self._init_device_mesh(mesh_shape=(2, 2), mesh_dim_names=("dp", "pp"))
+        pp_group = device_mesh["pp"].get_group()
+        dp_mesh = device_mesh["dp"]
+        assert type(pp_group) == dist.ProcessGroup
+        assert type(dp_mesh) == DeviceMesh
+
+        # create "entire model"
+        pp_group_size = pp_group.size()
+
+        # 8 layers
+        layers_per_model = 4
+        full_model = nn.ModuleList(
+            [nn.Linear(10, 10) for _ in range(pp_group_size * layers_per_model)]
+        )
+
+        # divide the model (8 layers) by the number of ranks (2)
+        partial_model = nn.Sequential(
+            *full_model[
+                pp_group.rank()
+                * layers_per_model : (pp_group.rank() + 1)
+                * layers_per_model
+            ]
+        )
+        partial_model.to(device)
+
+        # apply FSDP
+        mp_policy = MixedPrecisionPolicy(
+            # TODO(whc) need to fix PP + FSDP-mixed-precision
+            # tracer for PP assumes f32 and is caught off guard when runtime FSDP interacts using bf16 inputs
+            # param_dtype=torch.bfloat16, reduce_dtype=torch.float32
+            param_dtype=torch.float32,
+            reduce_dtype=torch.float32,
+        )
+        fsdp_config = {"mesh": dp_mesh, "mp_policy": mp_policy}
+        for layer_name, layer in partial_model.named_children():
+
+            # As an optimization, do not reshard after forward for the last
+            # transformer block since FSDP would prefetch it immediately
+            # reshard_after_forward = layer_id < len(model.layers) - 1
+            # TODO(whc) need to fix correctly handle layer-ids on pp-split module
+            fully_shard(
+                layer,
+                **fsdp_config,
+                reshard_after_forward=True,
+            )
+            setattr(partial_model, layer_name, layer)
+        fsdp_model = fully_shard(partial_model, **fsdp_config)
+
+        # apply PP
+        input1 = torch.rand((1, 10), device=device)
+        num_microbatches = 8
+        pipeline_stage = self._create_manual_pipeline_stage(
+            fsdp_model,
+            pp_group.rank(),
+            pp_group.size(),
+            device,
+            pp_group,
+            input1,
+            num_microbatches,
+        )
+
+        pipeline_schedule = ScheduleGPipe(
+            pipeline_stage,
+            n_microbatches=num_microbatches,
+            # dummy loss needed just to force backwards to run in schedule step
+            loss_fn=lambda y, t: y.sum(dim=1),
```

Comment:
im doing something wrong here? rank0,2 is complaining about shapes.  But loss should be running on rank1,3. so i'm confused.

RuntimeError:       
        Failed to run stage backward:
        Stage output: ('Tensor(torch.Size([10]), grad=True)',)
        Output gradient: ['Tensor(torch.Size([1, 10]), grad=False)']
        Input: ['Tensor(torch.Size([1, 10]), grad=False)']

Reply:
@H-Huang can you help please?

Reply:
https://github.com/pytorch/PiPPy/issues/1054
made it an issue.  and found a workaround but need to discuss what the right fix is.